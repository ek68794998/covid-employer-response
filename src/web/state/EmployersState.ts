import { EmployerRecord } from "../../common/EmployerRecord";

export interface EmployersState {
	items: EmployerRecord[];
}