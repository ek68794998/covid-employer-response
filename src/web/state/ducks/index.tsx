import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});