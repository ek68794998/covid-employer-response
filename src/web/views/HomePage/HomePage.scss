@import "../Colors.scss";
@import "../Dimensions.scss";

$body-vert-width: 920px;
$card-area-padding: 16px;
$cell-width: ($body-min-width - $card-area-padding * 2);

main#home {
	display: flex;
	flex-direction: column;
	margin: 0 auto;
	width: 100%;

	> section {
		padding: 0 $page-padding;

		&.HomePage__Primer {
			background: $header;
			color: $background;

			> div {
				margin: 0 auto;
				max-width: $body-max-width;
				padding: 0 $page-padding $page-padding;
				text-align: center;
	
				.HomePage__PreTitle {
					font-size: 30px;
					margin-bottom: -6px;
					opacity: 0.7;
				}
	
				h1 {
					font-size: 80px;
					text-transform: uppercase;
					margin: 0 0 24px;
					padding: 0;
				}
	
				h2 {
					font-size: 20px;
					font-weight: 400;
					margin: 0;
					opacity: 0.0;
					padding: 0;
	
					&.HomePage__Subtitle--Loaded {
						opacity: 1.0;
						transition: 2s opacity;
						transition-delay: 0.5s;
					}
	
					a {
						margin: 0 6px;
					}
				}
			}
		}

		&.HomePage__Sample {
			display: flex;
			flex-wrap: wrap;
			justify-content: center;
			margin: 0 auto;
			max-width: ($cell-width * 2 + $card-area-padding * 2 + $page-padding);
			padding: $page-padding $card-area-padding !important;
			width: 100%;

			> div.HomePage__SampleInfo {
				align-self: center;
				max-width: $cell-width;
				width: 50%;

				h3 {
					font-size: 32px;
					margin: 16px 0;
				}

				p {
					font-size: 16px;
					line-height: 200%;
					padding: 0;
				}
			}

			> div.HomePage__SampleSpacer {
				width: $page-padding;
			}

			> div.HomePage__SampleCard {
				max-width: $cell-width;
				width: 50%;

				> div {
					height: 100%;
				}
			}
		}

		&.HomePage__Links {
			background: $header;
			flex: 1;
			text-align: center;

			a {
				background: $header;
				border: 1px solid $outline-color;
				border-radius: 32px;;
				color: $outline-color;
				cursor: pointer;
				display: inline-block;
				font-family: inherit;
				font-size: 24px;
				height: 64px;
				margin: 30px auto;
				padding: 16px 48px;

				&:hover {
					background: $outline-color;
					color: $header;
				}
			}
		}
	}
}

@media (max-width: $body-vert-width) {
	main#home {
		> section {
			&.HomePage__Sample {
				> div.HomePage__SampleInfo, > div.HomePage__SampleCard {
					flex: 1;
				}

				> div.HomePage__SampleSpacer {
					flex-basis: 100%;
					height: $page-padding-narrow;
					width: 0;
				}
			}
		}
	}
}

@media (max-width: $body-narrow-width) {
	main#home {
		> section {
			padding: 0 $page-padding-narrow !important;

			&.HomePage__Primer > div {
				padding: 0 $page-padding-narrow $page-padding !important;
			}
		}
	}
}
