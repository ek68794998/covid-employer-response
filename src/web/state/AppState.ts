import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}