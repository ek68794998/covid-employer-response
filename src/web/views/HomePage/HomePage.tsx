import React, { useEffect, useState } from "react";
import ReactMarkdown from "react-markdown";
import { useSelector } from "react-redux";
import { Link, withRouter } from "react-router-dom";

import { format, LocalizedStrings } from "../../../common/LocalizedStrings";
import { ProjectIssueSubmissionUrl, SubmissionFormUrl } from "../../../common/constants/UrlConstants";

import { getEmployerIds } from "../../state/ducks/employers/selectors";
import { getStrings } from "../../state/ducks/localization/selectors";

import AppHelmet from "../AppHelmet/AppHelmet";
import BackToTopButton from "../BackToTopButton/BackToTopButton";
import EmployerListItem from "../EmployerListItem/EmployerListItem";

import "./HomePage.scss";

const HomePage: React.FC = (): React.ReactElement => {
	const strings: LocalizedStrings = useSelector(getStrings);

	const [ randomEmployerId, setRandomEmployerId ] = useState<string | undefined>(undefined);

	const employerIds: string[] | undefined = useSelector(getEmployerIds);
	const employerCount: number = employerIds ? employerIds.length : 0;

	useEffect(
		() => {
			setRandomEmployerId(
				employerIds && employerIds[Math.floor(Math.random() * employerIds.length)]);
		},
		[ employerIds?.length ]);

	return (
		<main id="home">
			<AppHelmet
				title={strings.home}
			/>
			<section className="HomePage__Row HomePage__Row--Alt HomePage__Primer">
				<div>
					<div className="HomePage__PreTitle">Welcome to</div>
					<h1>
						{strings.appTitleShort}
					</h1>
					<h2 className={`HomePage__Subtitle--${employerCount > 0 ? "Loaded" : "Loading"}`}>
						<ReactMarkdown
							allowedTypes={[ "text", "strong", "emphasis", "link" ]}
							source={format(strings.homeSubtitleText, { count: employerCount.toLocaleString() })}
							unwrapDisallowed={true}
						/>
					</h2>
				</div>
			</section>
			<section className="HomePage__Row">
				<div>
					<div className="HomePage__InfoBlock HomePage__TileBlock">
						<h3>{strings.homeSampleTitle}</h3>
						<ReactMarkdown source={strings.homeSampleDescription} />
					</div>
					<div className="HomePage__Spacer" />
					<div className="HomePage__SampleCard HomePage__TileBlock">
						{randomEmployerId && (
							<EmployerListItem
								employerId={randomEmployerId}
								metric="rating"
								showDetails={true}
							/>
						)}
					</div>
				</div>
			</section>
			<section className="HomePage__Row HomePage__Links">
				<Link className="App__BigButton" to="/employers">{strings.homeListLink}</Link>
			</section>
			<section className="HomePage__Row HomePage__Row--Alt HomePage__Row--Reverse HomePage__Submit">
				<div>
					<div className="HomePage__TileBlock">
						<img src={"/images/google-form.png"} />
					</div>
					<div className="HomePage__Spacer" />
					<div className="HomePage__InfoBlock HomePage__TileBlock">
						<h3>{strings.homeSubmitTitle}</h3>
						<ReactMarkdown
							source={(
								format(strings.homeSubmitDescription, { githubIssueUrl: ProjectIssueSubmissionUrl })
							)}
						/>
					</div>
				</div>
			</section>
			<section className="HomePage__Row HomePage__Row--Alt HomePage__Links">
				<a
					className="App__BigButton"
					href={SubmissionFormUrl}
					rel="noopener noreferrer"
					target="_blank"
				>
					{strings.homeFormLink}
				</a>
			</section>
			<BackToTopButton />
		</main>
	);
};

export default withRouter(HomePage);
