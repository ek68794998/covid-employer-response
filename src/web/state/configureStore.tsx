import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;