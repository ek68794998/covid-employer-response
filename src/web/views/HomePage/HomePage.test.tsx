import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { getPlocStringsAsync } from "../../../__tests__/TestUtils";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import HomePage from "./HomePage";

describe("<HomePage />", (): void => {
	test("renders without exploding", async (): Promise<void> => {
		const store: Store<AppState, AnyAction> = configureStore({ strings: await getPlocStringsAsync() });

		const renderedValue: ReactTestRenderer =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<HomePage />
					</BrowserRouter>
				</Provider>,
			);

		expect(renderedValue.toJSON()).toMatchSnapshot();
	});
});
