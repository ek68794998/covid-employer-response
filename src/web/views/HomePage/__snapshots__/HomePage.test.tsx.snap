// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> renders without exploding 1`] = `
<main
  id="home"
>
  <section
    className="HomePage__Primer"
  >
    <div>
      <div
        className="HomePage__PreTitle"
      >
        Welcome to
      </div>
      <h1>
        áƥƥ
      </h1>
      <h2
        className="HomePage__Subtitle--Loading"
      >
        A free, open-data collection on
        <a
          href="/employers"
          onClick={[Function]}
        >
          <strong>
            0
          </strong>
           employers
        </a>
        (and growing) and how they've reacted to the COVID-19 crisis.
      </h2>
    </div>
  </section>
  <section
    className="HomePage__Sample"
  >
    <div
      className="HomePage__SampleInfo"
    >
      <h3>
        What can I learn?
      </h3>
      <p>
        Each card shows you briefly 
        <strong>
          who
        </strong>
         the employer is, a 
        <strong>
          rating
        </strong>
         of their general response, and a 
        <strong>
          summary
        </strong>
         about the actions they've taken. You can also open up a more detailed view to find 
        <strong>
          cited news articles
        </strong>
        , or click various 
        <strong>
          links
        </strong>
         to go on Wikipedia, or see the employer's official website.
      </p>
    </div>
    <div
      className="HomePage__SampleSpacer"
    />
    <div
      className="HomePage__SampleCard"
    />
  </section>
  <section
    className="HomePage__Links"
  >
    <a
      href="/employers"
      onClick={[Function]}
    >
      Take me to the list!
    </a>
  </section>
</main>
`;
